### PR TITLE
Hbonds

### PR DESCRIPTION
**Pull request by lscalfi**
This is a command to print the hydrogen bond network of a system, frame by frame. It outputs a list of: donor hydrogen acceptor : distance d-a   angle d-a-h.
It is possible to select the donors and the acceptors possible (the acceptor selection should be of size 2).

### CODE REVIEW

**Comment by Luthaf** [comment]
Great to see this! Thanks for submitting the PR. 

I know we talked about this in person, but could you add a small description in the initial PR comment next time? This allow external readers to see what is going on.

File: src/commands/HBonds.cpp [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+#include "HBonds.hpp"
+#include "Errors.hpp"
+#include "utils.hpp"
+#include "geometry.hpp"
+
+using namespace chemfiles;
+
+static const double pi = 3.141592653589793238463;
+static const char OPTIONS[] =
+R"(Compute list of hydrogen bonds along a trajectory. A selection can be specified using the
```
Comment (Luthaf) [inline src/commands/HBonds.cpp]: "A selection can be specified .." => "Selections for the acceptor and donor atoms can be specified ..."

File: src/commands/HBonds.cpp [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+#include "geometry.hpp"
+
+using namespace chemfiles;
+
+static const double pi = 3.141592653589793238463;
+static const char OPTIONS[] =
+R"(Compute list of hydrogen bonds along a trajectory. A selection can be specified using the 
+chemfiles selection language. It is possible to provide an alternative unit cell or topology 
+for the trajectory file if they are not defined in the trajectory format. Hydrogen bonds criteria
+can be specified (donor-acceptor distance and donor-acceptor-H angle)
```
Comment (Luthaf) [inline src/commands/HBonds.cpp]: Maybe add an explanation/comment about how the hydrogen bonds are detected? (using geometrical criteria)

File: src/commands/HBonds.cpp [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+For more information about chemfiles selection language, please see
+http://chemfiles.github.io/chemfiles/latest/selections.html
```
Comment (Luthaf) [inline src/commands/HBonds.cpp]: can you change the URL to `http://chemfiles.org/chemfiles/...`?

File: src/commands/HBonds.cpp [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+Usage:
+  cfiles hbonds [options] <trajectory>
+  cfiles hbonds (-h | --help)
+
+Examples:
+  cfiles hbonds --cell=28 --guess-bonds water.xyz water.pdb
+  cfiles hbonds butane.pdb butane.nc --wrap
+  cfiles hbonds methane.xyz --cell 15:15:25 --guess-bonds --points=150
+  cfiles hbonds result.xtc --topology=initial.mol --topology-format=PDB out.nc
+  cfiles hbonds in.zeo out.mol --input-format=XYZ --output-format=PDB
```
Comment (Luthaf) [inline src/commands/HBonds.cpp]: These examples make no sense. copy/paste error ?

File: src/commands/HBonds.cpp [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+  -c <cell>, --cell=<cell>      alternative unit cell. <cell> format is one of
+                                <a:b:c:α:β:γ> or <a:b:c> or <a>. 'a', 'b' and
+                                'c' are in angstroms, 'α', 'β', and 'γ' are in
+                                degrees.
+  --steps=<steps>               steps to use from the input. <steps> format
+                                is <start>:<end>[:<stride>] with <start>, <end>
+                                and <stride> optional. Default is to use all
+                                steps from the input; starting at 0, ending at
+                                the last step, and with a stride of 1.
+  --selection_acceptor=<sel>    selection to use for the acceptors. This must be a
```
Comment (Luthaf) [inline src/commands/HBonds.cpp]: How about `--acceptors` for the argument name?

File: src/commands/HBonds.cpp [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+                                selection of size 2 and type bonds.
+                                [default: bonds: type(#2) == H]
+  --selection_donor=<sel>       selection to use for the donors. This must be a
```
Comment (Luthaf) [inline src/commands/HBonds.cpp]: Same here, how about `--donnors`

File: src/commands/HBonds.cpp [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+                                selection of size 1.
+                                [default: atoms: not type H]
```
Comment (Luthaf) [inline src/commands/HBonds.cpp]: Shouldn't the default be `type O or type N or type F `?

File: src/commands/HBonds.cpp [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+  -p <par>, --parameters=<par>  parameters to use for the hydrogen bond. <par>
```
Comment (Luthaf) [inline src/commands/HBonds.cpp]: I would use two separated parameters here: `--distance` and `--angle`

File: src/commands/HBonds.cpp [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+                                format is <d:α> where 'd' is the donor-acceptor 
+                                maximum distance in angstroms and 'α' is the
+                                donor-acceptor-hydrogen maximum angle in degrees.
+                                [default: 3.0:30.0]
+)";
+
+static HBonds::Options parse_options(int argc, const char* argv[]) {
+    auto options_str = command_header("hbonds", HBonds().description()) + "\n";
+    options_str += OPTIONS;
```
Comment (Luthaf) [inline src/commands/HBonds.cpp]: Can you add your author informations here? 

```cpp
auto options_str = command_header("hbonds", HBonds().description());
options_str += "Laura Scalfi <email>\n";
options_str += OPTIONS;
```

File: src/commands/HBonds.hpp [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        /// Input trajectory
+        std::string trajectory;
+	/// Output
+	std::string outfile;
+        /// Specific format to use with the trajectory
+        std::string format = "";
+        /// Specific steps to use from the trajectory
+        steps_range steps;
+	/// Selection for the donor-acceptor
+	std::string selectionAcceptor;
```
Comment (Luthaf) [inline src/commands/HBonds.hpp]: Can you use `snake_case` here?

File: src/commands/HBonds.hpp [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+	double angle_parameter = 0.0;
+    };
+
+    HBonds(): selectionAcceptor_("bonds: type(#2) == H"), selectionDonor_("not type H") {}
+    int run(int argc, const char* argv[]) override;
+    std::string description() const override;
+
+private:
+    Options options_;
+    chemfiles::Selection selectionAcceptor_;
```
Comment (Luthaf) [inline src/commands/HBonds.hpp]: Can you use `snake_case` here too?

File: src/commands/HBonds.hpp [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        std::string topology_format = "";
+        /// Should we try to guess the topology?
+        bool guess_bonds = false;
+	/// Parameters for donor-acceptor max distance (in angstroms)
+	double distance_parameter = 0.0;
+	/// and for donor-acceptor-hydrogen max angle (in degrees)
+	double angle_parameter = 0.0;
+    };
+
+    HBonds(): selectionAcceptor_("bonds: type(#2) == H"), selectionDonor_("not type H") {}
```
Comment (Luthaf) [inline src/commands/HBonds.hpp]: Please do not provide defaults both here and in the doctopt `OPTIONS` string. There should be only one source of information to keep it [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

File: src/commands/HBonds.cpp [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+    options.trajectory = args.at("<trajectory>").asString();
+    options.guess_bonds = args.at("--guess-bonds").asBool();
+
+    options.selectionAcceptor = args.at("--selection_acceptor").asString();
+    options.selectionDonor = args.at("--selection_donor").asString();
+
+    if (args.at("--output")){
+        options.outfile = args.at("--output").asString();
+    } else {
+        options.outfile = options.trajectory + "_hb.dat";
```
Comment (Luthaf) [inline src/commands/HBonds.cpp:90]: This does not match the documentation.

File: src/commands/HBonds.cpp [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+        if (options.topology == "") {
+            throw CFilesError("Useless '--topology-format' without '--topology'");
+        }
+        options.topology_format = args.at("--topology-format").asString();
+    }
+
+    if (args.at("--cell")) {
+        options.custom_cell = true;
+	options.cell = parse_cell(args.at("--cell").asString());
+	}
```
Comment (Luthaf) [inline src/commands/HBonds.cpp]: Please use 4-spaces for indent (here and two lines below)

File: src/commands/HBonds.cpp [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+
+    options.distance_parameter = 3.0;
+    options.angle_parameter = 30.0;
+    if (args.at("--parameters")) {
+	auto splitted = split(args.at("--parameters").asString(), ':');
+	if (splitted.size() == 2) {
+	    options.distance_parameter = stod(splitted[0]);
+	    options.angle_parameter = stod(splitted[1])*pi/180;
```
Comment (Luthaf) [inline src/commands/HBonds.cpp]: What happens if `splitted[0] == "hviousbo"` ? (look for `std::invalid_argument` [here](http://en.cppreference.com/w/cpp/string/basic_string/stof))

File: src/commands/HBonds.cpp [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+std::string HBonds::description() const {
+    return "compute hydrogen bonds network";
+}
+
+int HBonds::run(int argc, const char* argv[]) {
+    auto options = parse_options(argc, argv);
+
+    selectionAcceptor_ = Selection(options.selectionAcceptor);
+    if (selectionAcceptor_.size() != 2) {
+        throw CFilesError("Can not use a selection for acceptors with size different than 2.");
```
Comment (Luthaf) [inline src/commands/HBonds.cpp]: You require `bonds:` selection in the documentation, but the selection could be of kind `pairs:` too here.

File: src/commands/HBonds.cpp [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+
+        auto positions = frame.positions();
+        auto cell = frame.cell();
+
+        auto matched = selectionAcceptor_.evaluate(frame);
+        for (auto match: matched) {
+            assert(match.size() == 2);
+
+            auto acceptor = match[0];
+            auto hydrogen = match[1];
```
Comment (Luthaf) [inline src/commands/HBonds.cpp]: What happens if the user provides acceptor selection like this: `bonds: name(#1) H and name(#2) O`?

File: src/commands/HBonds.cpp [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+            auto donors = selectionDonor_.list(frame);
+            for (auto donor: donors) {
+                if (donor != acceptor && frame.topology()[donor].type() != "H") {
+                    auto r_ad = cell.wrap(positions[donor] - positions[acceptor]);
+                    auto distance = norm(r_ad); 
+                    auto r_ah = cell.wrap(positions[hydrogen] - positions[acceptor]);
+                    auto theta = angle(r_ad, r_ah);
+                    if (distance < options.distance_parameter && theta < options.angle_parameter) {
+                        outfile << frame.topology()[donor].type() << donor << "   " << frame.topology()[acceptor].type() << acceptor << "   " << frame.topology()[hydrogen].type() << hydrogen << "  : " << distance << "    " << theta*180/pi << std::endl;
```
Comment (Luthaf) [inline src/commands/HBonds.cpp]: could you split this on multiple lines? It is really hard to read.

**Review by Luthaf — CHANGES_REQUESTED** [review]
Great work for a first time with C++! Here are a lot of comments, but don't be afraid, all of them are easy to change.

We also need to think a bit more about the output format. Currently it is not self describing (we need a comment telling what is in each column of the output); and multiple timesteps are not separated.

File: src/commands/HBonds.cpp (hunk 4 above)
Comment (lscalfi) [reply src/commands/HBonds.cpp]: oups!

File: src/commands/HBonds.cpp [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+http://chemfiles.org/chemfiles/latest/selections.html
+
+Usage:
+  cfiles hbonds [options] <trajectory>
+  cfiles hbonds (-h | --help)
+
+Examples:
+  cfiles hbonds water.xyz --cell 15:15:25 --guess-bonds 
+  cfiles hbonds in.pdb --acceptors=="bonds: type(#1) == O and type(#2) == H"
+  cfiles hbonds protein.pdb --donors=="atoms: type N" -p 2.5:20.0
```
Comment (Luthaf) [inline src/commands/HBonds.cpp]: The syntax for parameters changed =)

File: src/commands/HBonds.cpp [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+
+            if (frame.topology()[match[0]].type() == "H") {
+                acceptor = match[1];
+                hydrogen = match[0];
+            }
+            else if (frame.topology()[match[1]].type() != "H") {
+                acceptor = match[0];
+                hydrogen = match[1];
+            }
+            else
```
Comment (Luthaf) [inline src/commands/HBonds.cpp]: Could you write this `} else {` in a single line to follow the rest of the code ?

File: src/commands/HBonds.cpp [hunk 21]
```diff
@@ -0,0 +1,238 @@
...
+
+    auto infile = Trajectory(options.trajectory, 'r', options.format);
+    std::ofstream outfile(options.outfile, std::ios::out);
+    if (outfile.is_open()) {
+        outfile << "#Hydrogen bond network in trajectory " << options.trajectory << std::endl;
+        outfile << "# Selection: acceptors: " << options.selection_acceptor;
+        outfile << " and donors: " << options.selection_donor << std::endl;
+        outfile << "# Criteria:" << std::endl;
+        outfile << "# donor-acceptor distance < " << options.distance << " angstroms" << std::endl;
+        outfile << "# donor-acceptor-H angle < " << options.angle*180/pi << " degrees" << std::endl;
```
Comment (Luthaf) [inline src/commands/HBonds.cpp:171]: This still misses a comment line describing the content of the file.

File: src/commands/HBonds.cpp [hunk 22]
```diff
@@ -0,0 +1,238 @@
...
+
+            auto donors = selection_donor_.list(frame);
+            for (auto donor: donors) {
+                if (donor != acceptor && frame.topology()[donor].type() != "H") {
+                    auto r_ad = cell.wrap(positions[donor] - positions[acceptor]);
+                    auto distance = norm(r_ad); 
+                    auto r_ah = cell.wrap(positions[hydrogen] - positions[acceptor]);
+                    auto theta = angle(r_ad, r_ah);
+                    if (distance < options.distance && theta < options.angle) {
+                        outfile << frame.topology()[donor].type() << donor << "   ";
```
Comment (Luthaf) [inline src/commands/HBonds.cpp]: Why use `type` instead of `name` here?

**Comment by Luthaf** [comment]
Can you add a test to ensure the coverage stays at least at the same level? 

This should be the last thing to do 😃 !

File: src/commands/HBonds.cpp [hunk 23]
```diff
@@ -191,7 +191,7 @@ int HBonds::run(int argc, const char* argv[]) {
         }
 
         outfile << "# Frame: " << step << std::endl;        
-        outfile << "# Donor   Acceptor   Hydrogen  : Distance D-A    Angle D-A-H" << std::endl;
+        outfile << "# Donor (name index)   Acceptor (name index)   Hydrogen (name index)  : Distance D-A    Angle D-A-H" << std::endl;
```
Comment (Luthaf) [inline src/commands/HBonds.cpp]: Shouldn't this angle be D-H-A ?

Comment (lscalfi) [reply src/commands/HBonds.cpp]: no, imagine this configuration:
..................... D
   A-----H  
We want A, H and D to be the "most co-linear" possible. If the angle D-A-H is zero, then they are aligned, if not <e want the angle to be less than 30°.
By taking D-H-A, we would want it to be equal to 180° +/- 30°.

**Review by Luthaf — APPROVED** [review]

**Comment by Luthaf** [comment]
Thank you again for this!

**Pull request merged**